### PR TITLE
Activation offloading to CPU's for the Linear, Layernorm Linear and the Layernorm MLP modules

### PR DESCRIPTION
**Pull request by sanandaraj5597**
This PR adds support for offloading all the tensors saved for backward pass by the Linear, Layernorm Linear and the Layernorm MLP modules except the weight tensors.

The _cpu_offloading_ switch will be sent from the M-LM module construction which will enable calling the PyTorch hooks during tensor saving and retrieval. When these PyT hooks are called, the weight.main_grad isn't saved, so we save them separately on a need basis when we need to fuse gradients. All these layer executions will be done under a context which is placed at M-LM repo.

Please review and let me know if you have any questions.

### CODE REVIEW

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:1]: Please add the license and copyright information.

File: transformer_engine/pytorch/module/layernorm_linear.py
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm_linear.py:73]: Hmmm, I'm not 100% sure I understand the interaction between this option and the context manager. Naively, seeing this option I would expect the context manager to not do anything if this option is not set as well. When looking at the code though, it seems to mostly set the tensors that we do NOT want to offload. Could you explain that?

Additionally, could you comment on whether we expect that granular, module by module control of the offloading to be needed? Maybe we can just check for the state of the context manager to figure out whether we should offload instead?

Comment (sanandaraj5597) [reply transformer_engine/pytorch/module/layernorm_linear.py:73]: >> When looking at the code though, it seems to mostly set the tensors that we do NOT want to offload. Could you explain that?
You are right, the context manager doesn't do anything. The problem occurs when cpu_offload is enabled when fuse_wgrad_accumulation is also enabled. To give you some background, when we use the context manager to do something when we save/restore tensors, Torch removes the child tensors of the Parameter we try to save. To be precise, weight Parameter loses its main_grad when saved under this cpu offload context. That's why in backward function I reconstruct the Parameter object with main_grad that we are saving. You may ask me why can't we do this generically agnostic to fuse_wgrad_accumulation and cpu_offload, the problem is when cpu_offload is true and fuse_wgrad_accumulation is False, the autograd torch engine sees the weight object being over ridden with a new object and fails. So we have few combinations of cpu_offload and fuse_wgrad_accumulation based on which we have to decide to save main_grad in forward pass and reconstruct the weight Parameter object in backward class.

 >>Additionally, could you comment on whether we expect that granular, module by module control of the offloading to be needed?
The only reason I see is that each module has different type of weights, and we need to selectively offload/avoid offload them based on each module. The above response also adds up to the response of the question.

>>Maybe we can just check for the state of the context manager to figure out whether we should offload instead?
I don't understand what you mean by this, if the above explanation doesn't suffice, please elaborate on this so that we can discuss.

Comment (ptrendx) [reply transformer_engine/pytorch/module/layernorm_linear.py:73]: Hmmm, I think there is some confusion here so let me rephrase my questions then.

Basically, the way I see this PR right now is that the user needs to do 2 things when they want to use cpu offloading - they need to use the context manager and set the `cpu_offloading` flag to True. The problem is that setting only 1 of these things does not seem to do the right thing - using just the context manager will break the wgrad accumulation fusion, while setting just the option in the module will not really do anything. What is more, the existence of this `cpu_offloading` parameter gives a (I believe) false impression that there is a granular control over the feature - as a user I would assume that I can set the `cpu_offloading` flag to `False` in order to exclude a particular module from using the offload, which I believe is not actually true.

Therefore it seems to me that this is actually a binary switch (either both context manager and the option in the module are set or neither of them is set), so why not combine the 2 together? Basically inside the modules we can check if we are inside the CPU offloading context manager and if so, behave as if the `cpu_offloading` parameter was set to True. We can also pass that information to backward (e.g. via `ctx` in the autograd functions if we want to have the ability to turn it on and off dynamically or in the `self.cpu_offloading` if we believe it is a feature that will not be switched at will during training) in order to not have to call backward inside the context manager.

Does that make sense now?

File: transformer_engine/pytorch/module/linear.py
Comment (ptrendx) [inline transformer_engine/pytorch/module/linear.py:595]: This is not consistent with the actual parameters passed to __init__.

What I would do here is something like this:
in `cpu_offloading.py`:
```python
CPUOffloadEnabled = False

class CPUOffloadSaveTensorHook:
    def __enter__():
        global CPUOffloadEnabled
        CPUOffloadEnabled = True
        # the rest of the logic
    def __exit__():
        global CPUOffloadEnabled
        CPUOffloadEnabled = False
        # the rest of the logic
```
then in each module you would not have any parameter, but it would just check the state during `__init__`:
```python
self.cpu_offloading = CPUOffloadEnabled
```

**Comment by ptrendx** [comment]
Could you add some unit test for the functionality?

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:10]: nit: Context-manager

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:377]: Please add the docs to this function (as I believe this to be the user-facing function, right?) and add it to the autodoc here: https://github.com/NVIDIA/TransformerEngine/blob/main/docs/api/pytorch.rst

Please also add it to the __init__.py in transformer_engine/pytorch so that it is importable by
```python
from transformer_engine.pytorch import get_cpu_offload_context
```

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:8]: Please add
```python
__all__ = ['get_cpu_offload_context']
```
and pass there only the function we intend people to use (I think it is just this 1 function, but maybe there are more?)

File: transformer_engine/pytorch/module/linear.py
Comment (sanandaraj5597) [reply transformer_engine/pytorch/module/linear.py:595]: We discussed to pass ContextManager to __init__ of the modules earlier right, am I missing something? 

I can change the description to match the ContextManager expectations.

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:79]: nit: Context-manager

File: transformer_engine/pytorch/cpu_offload.py
Comment (ptrendx) [inline transformer_engine/pytorch/cpu_offload.py:105]: nit: what 2 methods? I would probably just say "A base class for CPU offload-handler"

**Review by ptrendx — APPROVED** [review]
Waiting for unit test and then lgtm.

File: transformer_engine/pytorch/cpu_offload.py
Comment (ksivaman) [inline transformer_engine/pytorch/cpu_offload.py:88]: This seems to be a general functionality instead of it being specific to CPU offloading, could we rename it as such?

Comment (ksivaman) [reply transformer_engine/pytorch/cpu_offload.py:88]: Additionally, is this a reimplementation of `torch.autograd.graph.saved_tensors_hooks`?

**Review by ksivaman — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/cpu_offload.py
Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py:88]: >>Additionally, is this a reimplementation of torch.autograd.graph.saved_tensors_hooks?

Yes, this is a modified version of the hook you mentioned but not a re-implementation. We wanted to maintain state across different calls to this hook, so we made it as a Class object.

Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py:88]: >This seems to be a general functionality instead of it being specific to CPU offloading, could we rename it as such?

I don't understand how do you call it generic, because this Class needs to get a offload handler based on which offloading happens. The offload handler handles asynchronously launching CUDA streams. For forward, it offloads activations/weights of the previously computed TF layer while for backward it reloads the next layers weight/activations. 

The offload handler provides tensor_pop/tensor_push function which is used inside the hook's on_save_for_backward and on_get_saved_tensor functions which are called during the context's enter function.
The synchronizer provides a sync barrier until both the offloading/reloading and the computation completes.

Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py:88]: @ksivaman : Can we resolve this convo? Do you have any feedback?

**Comment by ksivaman** [comment]
I've fixed most of the functional and lint errors and added some new tests as well as opposed to the singular test file that wasn't being used. The offloading for the `TransformerLayer` is failing with the following error due to the `LayerNormMLP` block, specifically during the offloading of weights.

````
RuntimeError: Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()
````

**Comment by ksivaman** [comment]
@sanandaraj5597 The above bug is a result of an attempted offload of a `[4, 4]` tensor, which seems unusual.
Additionally, the `Linear` and `LayerNormLinear` modules are not setting the `weight_offloading` attribute for the `weight` itself.

**Comment by sanandaraj5597** [comment]
Fixed the issue you were seeing Kirthi. Please review. Thank you.

**Review by ksivaman — APPROVED** [review]
1. FP8 cases were failing since the intermediate buffers for offloading and copying to the CPU were not `Float8Tensor` compatible. 
2. The inputs saved could be null tensors, thus assigning the `activation_offloading` attribute directly caused some issues.
3. LayerNorm and RMSNorm should also both work now.

LGTM, CI pending.

**Pull request merged**
